src/tools/chip-cert/CertUtils.cpp

```diff
@@ -0,0 +1,826 @@
+
+    VerifyOrExit(certLen <= INT_MAX, res = false);
+
+    certBIO = BIO_new_mem_buf(reinterpret_cast<void *>(cert), static_cast<int>(certLen));
+    if (certBIO == nullptr)
+    {
+        fprintf(stderr, "Memory allocation error\n");
+        ExitNow(res = false);
+    }
+
+    if (!PEM_read_bio(certBIO, &name, &header, &data, &dataLen))
+    {
+        ReportOpenSSLErrorAndExit("PEM_read_bio", res = false);
+    }
+
+    VerifyOrExit((dataLen >= 0) && (dataLen <= UINT32_MAX), res = false);
+
+    memcpy(cert, data, static_cast<uint32_t>(dataLen));
+    certLen = static_cast<uint32_t>(dataLen);
+
+exit:
+    if (certBIO != nullptr)
+    {
+        BIO_free(certBIO);
+    }
+    if (name != nullptr)
+    {
+        OPENSSL_free(name);
+    }
+    if (header != nullptr)
+    {
+        OPENSSL_free(header);
+    }
+    if (data != nullptr)
+    {
+        OPENSSL_free(data);
+    }
+
+    return res;
+}
+
+static CertFormat DetectCertFormat(uint8_t * cert, uint32_t certLen)
+{
+    static const uint8_t chipRawPrefix[]   = { 0xD5, 0x00, 0x00, 0x08, 0x00, 0x01, 0x00 };
+    static const char * chipB64Prefix      = "1QAACAAB";
+    static const uint32_t chipB64PrefixLen = sizeof(chipB64Prefix) - 1;
+    static const char * pemMarker          = "-----BEGIN CERTIFICATE-----";
+
+    if (certLen > sizeof(chipRawPrefix) && memcmp(cert, chipRawPrefix, sizeof(chipRawPrefix)) == 0)
+    {
+        return kCertFormat_Chip_Raw;
+    }
+
+    if (certLen > chipB64PrefixLen && memcmp(cert, chipB64Prefix, chipB64PrefixLen) == 0)
+    {
+        return kCertFormat_Chip_Base64;
+    }
+
+    if (ContainsPEMMarker(pemMarker, cert, certLen))
+    {
+        return kCertFormat_X509_PEM;
+    }
+
+    return kCertFormat_X509_DER;
+}
+
+bool ReadCert(const char * fileName, X509 *& cert)
+{
+    CertFormat origCertFmt;
+    return ReadCert(fileName, cert, origCertFmt);
+}
+
+bool ReadCert(const char * fileName, X509 *& cert, CertFormat & origCertFmt)
+{
+    bool res = true;
+    CHIP_ERROR err;
+    uint8_t * certBuf = nullptr;
+    const uint8_t * p;
+    uint32_t certLen;
+    CertFormat certFmt;
+
+    cert = nullptr;
+
+    res = ReadFileIntoMem(fileName, certBuf, certLen);
+    VerifyTrueOrExit(res);
+
+    VerifyOrExit(certLen <= MAX_CERT_SIZE, res = false);
+
+    certFmt = origCertFmt = DetectCertFormat(certBuf, certLen);
+
+    if (certFmt == kCertFormat_X509_PEM)
+    {
+        res = X509PEMToDER(certBuf, certLen);
+        VerifyTrueOrExit(res);
+
+        certFmt = kCertFormat_X509_DER;
+    }
+    else if (certFmt == kCertFormat_Chip_Base64)
+    {
+        if (!Base64Decode(certBuf, certLen, certBuf, certLen, certLen))
+        {
+            ExitNow(res = false);
+        }
+        certFmt = kCertFormat_Chip_Raw;
+    }
+
+    if (certFmt == kCertFormat_Chip_Raw)
+    {
+        uint32_t convertedCertLen  = certLen * kMaxChipCertInflationFactor;
+        uint8_t * convertedCertBuf = reinterpret_cast<uint8_t *>(malloc(convertedCertLen));
```

Comment:
Please use `std::vector<uint8_t>` or `std::unique_ptr<uint8_t[]>` here and remove all the manual memory management with malloc/free.

Reply:
Updated everywhere in this PR.